### DISCUSSION TITLE
Moving a Claim to Another Namespace

### DISCUSSION THREAD

**Discussion by turkenh** [crossplane/crossplane#4081] [Knowledge base]
To move a claim to another namespace, you need to follow the following steps:

1. Get a copy of the claim you want to move to another namespace.
2. Change the `metadata.namespace` to the new one and apply it.
3. Update the references in the corresponding composite:
    - `metadata.labels[crossplane.io/claim-namespace]`
    - `spec.claimRef.namespace`
4. Delete the original claim.
5. Edit the original claim and remove the `spec.resourceRef` field. (or you could remove the finalizers).

### Example 

See the following as an example, where we will move the claim for [Cluster](https://github.com/upbound/platform-ref-gcp/blob/main/examples/cluster-claim.yaml) of [platform-ref-gcp](https://marketplace.upbound.io/configurations/upbound/platform-ref-gcp/v0.3.0) from the `default` namespace to `newns` namespace:

Initial State:

```bash
❯ kubectl get cluster.gcp.platformref -n default
NAME                       SYNCED   READY   CONNECTION-SECRET   AGE
platform-ref-gcp-cluster   True     True    cluster-conn        137m

# Corresponding composite
❯ kubectl get xcluster.gcp.platformref
NAME                             SYNCED   READY   COMPOSITION                            AGE
platform-ref-gcp-cluster-2zxd5   True     True    xclusters.gcp.platformref.upbound.io   137m
```

1. Get a copy of the claim you want to move to another namespace.

```bash
kubectl get cluster.gcp.platformref -n default platform-ref-gcp-cluster -o yaml > claim-orig.yaml
```

2. Change the `metadata.namespace` to the new one and apply it.

```diff
   name: platform-ref-gcp-cluster
-  namespace: default
+  namespace: newns
   resourceVersion: "3136"
   uid: 213f48f0-96bb-4bcd-909b-2ec09293c9e3
 spec:
```

```bash
kubectl apply -f claim-newns.yaml
```

Now, we have the two claims pointing the same composite but the new one fails with `cannot configure composite resource: cannot bind composite resource that references a different claim`.

```bash
❯ kubectl get claim --all-namespaces
NAMESPACE   NAME                                                          SYNCED   READY   CONNECTION-SECRET   AGE
default     cluster.gcp.platformref.upbound.io/platform-ref-gcp-cluster   True     True    cluster-conn        142m
newns       cluster.gcp.platformref.upbound.io/platform-ref-gcp-cluster   False            cluster-conn        15s
```

3. Update the references in the corresponding composite:
    - `metadata.labels[crossplane.io/claim-namespace]`
    - `spec.claimRef.namespace`

```bash
kubectl edit xclusters.gcp.platformref.upbound.io platform-ref-gcp-cluster-2zxd5
```

Now, when you check the claims, you will notice the new claim is Synced, but the old one is failing with the previous error: `cannot configure composite resource: cannot bind composite resource that references a different claim`.

```bash
❯ kubectl get claim --all-namespaces

NAMESPACE   NAME                                                          SYNCED   READY   CONNECTION-SECRET   AGE
default     cluster.gcp.platformref.upbound.io/platform-ref-gcp-cluster   False    True    cluster-conn        144m
newns       cluster.gcp.platformref.upbound.io/platform-ref-gcp-cluster   True     True    cluster-conn        2m49s
```

4. Initiate the deletion of the original claim (which will not complete because it's still in use).

```bash
kubectl delete cluster.gcp.platformref.upbound.io/platform-ref-gcp-cluster -n default
```

5. Edit the original claim to complete its deletion by removing the `spec.resourceRef` field. (or you could remove the finalizers instead).

```bash
kubectl edit cluster.gcp.platformref.upbound.io/platform-ref-gcp-cluster -n default
```

**Comment by jeanduplessis** [comment]
@plumbis how would you feel about adding this information to https://docs.crossplane.io/knowledge-base/ ?

**Reply by turkenh** [reply]
I don't think it is a common case that is worth including there. More aligned with other [KB discussions](https://github.com/crossplane/crossplane/discussions/categories/knowledge-base) here.

**Reply by jeanduplessis** [reply]
My concern is around the discoverability of this information. Maybe we can add something to that page that makes folks aware of the information here.

**Reply by plumbis** [reply]
I'm with Jean. Putting it in the docs will make it much easier for folks to find, both through Google and docs search.

**Comment by jbw976** [comment]
> 4. Delete the original claim
> 5. Edit the original claim

this confused me at first because intuitively, how would you edit something that's deleted? but the intent here is to make the deletion from step 4 actually complete.  Maybe updating the text like the below will help here?

4. Initiate the deletion of the original claim (which will not complete because it's still in use)
5. Edit the original claim to complete its deletion by removing the spec.resourceRef field. (or you could remove the finalizers instead).

**Reply by turkenh** [reply]
Thanks @jbw976!
Updated as you suggested.
